compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -10563,69 +10558,53 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
-            else
-               regDeps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(deps, 0, numDeps, cg);
-            }
-         else
-            {
-            if (needResultReg)
-               {
-               numDeps++; // for retValReg
-               regDeps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, numDeps, cg);
-               regDeps->addPostCondition(retValReg, TR::RealRegister::AssignAny);
-               }
-            }
-
-         if (length > (earlyCLCEnabled ? earlyCLCThreshold : TR_MAX_CLC_SIZE))
+         if (needResultReg)
             {
-            if (baseSource1Ref->getIndexRegister())
-               regDeps->addPostConditionIfNotAlreadyInserted(baseSource1Ref->getIndexRegister(), TR::RealRegister::AssignAny);
-
-            if (baseSource1Ref->getBaseRegister())
-               regDeps->addPostConditionIfNotAlreadyInserted(baseSource1Ref->getBaseRegister(), TR::RealRegister::AssignAny);
-
-            if (baseSource2Ref->getIndexRegister() && (baseSource2Ref->getIndexRegister() != baseSource1Ref->getIndexRegister()))
-               regDeps->addPostConditionIfNotAlreadyInserted(baseSource2Ref->getIndexRegister(), TR::RealRegister::AssignAny);
-
-            if (baseSource2Ref->getBaseRegister() && (baseSource2Ref->getBaseRegister() != baseSource1Ref->getBaseRegister()))
-               regDeps->addPostConditionIfNotAlreadyInserted(baseSource2Ref->getBaseRegister(), TR::RealRegister::AssignAny);
+            numDeps++; // for retValReg
+            regDeps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, numDeps, cg);
+            regDeps->addPostCondition(retValReg, TR::RealRegister::AssignAny);
+            generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, retValReg, retValReg);
             }
          }
 
-      TR::LabelSymbol *condLabel = isCondLabelNeeded ? generateLabelSymbol(cg) : NULL;
+      if (baseSource1Ref->getBaseRegister())
+         regDeps->addPostConditionIfNotAlreadyInserted(baseSource1Ref->getBaseRegister(), TR::RealRegister::AssignAny);
 
-      TR::LabelSymbol *endLabel = isEndLabelNeeded ? generateLabelSymbol(cg) : NULL;
+      if (baseSource2Ref->getBaseRegister() && (baseSource2Ref->getBaseRegister() != baseSource1Ref->getBaseRegister()))
+         regDeps->addPostConditionIfNotAlreadyInserted(baseSource2Ref->getBaseRegister(), TR::RealRegister::AssignAny);
+
+      //short circuit
+      TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+      TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+      TR::LabelSymbol *condLabel = generateLabelSymbol(cg);
+      generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionStart);
+      cFlowRegionStart->setStartInternalControlFlow();
 
-      if (!isFoldedIf && needResultReg)
-         generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, retValReg, retValReg);
+      generateS390CompareAndBranchInstruction(
```

Comment:
Jump if base registers are equal. offsets can not be different because of the line 10535.

Reply:
Do we need condition code from this comparison afterwards? If not, then can you set that parameter to `false`. Looking at the instruction selection log, you pasted, I do see that we generate CGR and BRC while we should be able to use `CGRJ` instruction (As both SRC1 and SRC2 is already in a register)

Reply:
done